keps/sig-apps/2214-indexed-job/README.md

```diff
@@ -0,0 +1,577 @@
+- All-or-nothing scheduling.
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a Job author, I can create an array Job where each Pod receives an ordered
+completion index. I can use the index in my binary through an environment
+variable or a file to statically select the load the Pod should work on.
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: parallel-work
+spec:
+  completions: 100
+  parallelism: 100
+  template:
+    spec:
+      containers:
+      - name: task
+        image: registry.example.com/processing-image
+        command: ["./process",  "--index", "$INDEX"]
+        env:
+        - name: INDEX
+          valueFrom:
+            fieldRef:
+              fieldPath: metadata.annotations['batch.alpha.kubernetes.io/job-completion-index'] 
+```
+
+### Notes/Constraints/Caveats (Optional)
+
+* An earlier proposal for [indexed Job]
+suggested the support for work lists, i.e. passing different parameters to each
+Pod. We decided to leave this out of the proposal to keep it simple and
+because work lists can be implemented in a startup script using the completion
+index as building block.
+* The semantics of an indexed Job are similar to a StatefulSet, in the sense
+that Pods have an associated index.
+However, the APIs have a major difference: a StatefulSet doesn't have completion
+semantics, as opposed to Jobs.
+
+[indexed Job]: https://github.com/kubernetes/community/blob/b21d1b27c8c748bf81283c2d89cde2becb5f2709/contributors/design-proposals/apps/indexed-job.md
+
+### Risks and Mitigations
+
+Jobs have a known issue in which more than one Pod can be started even if
+parallelism and completion are set to 1 ([reference]). In the case of indexed
+Jobs, this translates to more than one Pod having the same index.
+
+Just like for existing Job patterns, workloads have to handle duplicates at the
+application level.
+
+[reference]: https://kubernetes.io/docs/concepts/workloads/controllers/job/#handling-pod-and-container-failures
+
+## Design Details
+
+### JobSpec API
+
+The JobSpec gets the field `completionMode` to control whether the Job should	
+be treated as an Indexed Job.	
+
+```golang	
+type JobSpec struct {	
+  ...	
+  // CompletionMode specifies how Pod completions are tracked. It can be	
+  // `NonIndexed` (default) or `Indexed`.	
+  // `NonIndexed` means that each Pod completion is homologous to each other.	
+  // The Job is considered complete when there have been .spec.completions	
+  // successful completions.	
+  // `Indexed` means that each Pod completion needs to be tracked individually;	
+  // each Pods gets an associated completion index, which is available in the
+  // annotation	`batch.alpha.kubernetes.io/job-completion-index`.	
+  // The Job is considered complete when there is one successful Pod for each	
+  // index in the range 0 to (.spec.completions - 1).	
+  // When value is `Indexed`, .spec.completions must have a value between 1
+  // and 10^6.
```

Comment:
>  Yes, this is about being able to maintain the bitmask. Although, if we are going to use it to track completion, we actually need to cap .spec.completions too. The idea is that in the future we don't depend on lingering pods. @erictune thoughts?

I don't think you need to limit completions. The whole points that Eric made in one of other comments proved that it's enough to limit `parallelism` [because if you won't be storing bitmask, but rather [a-b,c-d,e-f,...] format, there will be at most "parallelism` of such pairs, so the length of that is actually bounded by parallelism.
Bounding completions is much more limiting that limiting parallelism to something as high as 10^6.

> Should we be doing some kind of client side throttling? In the current design, Job status updates would come after removing finalizers from finishing Pods, which already throttles Job updates in a way. But if one Pod finishes every 100ms (for example), we might still do more Job updates than what you are suggesting.

Yes - we should batch those, and send 1 update per X second. We can tune X later.

Reply:
> I don't think you need to limit completions. The whole points that Eric made in one of other comments proved that it's enough to limit parallelism [because if you won't be storing bitmask, but rather [a-b,c-d,e-f,...] format, there will be at most "parallelism` of such pairs, so the length of that is actually bounded by parallelism.

So it depends on the implementation. If we use a bitmask, then we have to limit completions. If we limit parallelism, we would have to use the compressed list format, but then the limit might have to be lower than 10^6, as we require more characters. We don't have to be concerned about display space, as we can just limit `kubectl describe` to print less characters.

Reply:
@soltysh any thoughts?

Reply:
Running 10^6 of pods at the same time is far from supported at this point.
Having more than 10^6 of pods within a single job sounds like a valid usecase to me.

So I would be much more comfortable saying that  "parallelism <= 10^5" (so decreasing that by order of magnitude even), than by setting a limit on the `completions`

Reply:
Moved back to limit parallelism and use a compressed list format instead of bitmask.